src/cache.rs

```diff
@@ -100,124 +138,167 @@ pub fn id() -> usize {
     {
-        self.map
+        self.entries
             .entry(key)
             .or_default()
-            .push(CacheEntry::new::<In, Out>(constraint, output));
+            .push(CacheEntry::new::<In>(constraint, output));
     }
 }
 
 /// A memoized result.
-struct CacheEntry {
-    /// The memoized function's constrained output.
-    ///
-    /// This is of type `Constrained<In::Constraint, Out>`.
-    constrained: Box<dyn Any>,
-    /// How many evictions have passed since the entry has been last used.
-    age: usize,
-}
-
-/// A value with a constraint.
-struct Constrained<C, T> {
-    /// The constraint which must be fulfilled for the output to be used.
+struct CacheEntry<C, Out> {
+    /// The memoized function's constraint.
     constraint: C,
     /// The memoized function's output.
-    output: T,
+    output: Out,
+    /// How many evictions have passed since the entry has been last used.
+    age: AtomicUsize,
 }
 
-impl CacheEntry {
+impl<C, Out: 'static> CacheEntry<C, Out> {
     /// Create a new entry.
-    fn new<In, Out>(constraint: In::Constraint, output: Out) -> Self
+    fn new<In>(constraint: In::Constraint, output: Out) -> Self
     where
-        In: Input,
-        Out: 'static,
+        In: Input<Constraint = C>,
     {
-        Self {
-            constrained: Box::new(Constrained { constraint, output }),
-            age: 0,
-        }
+        Self { constraint, output, age: AtomicUsize::new(0) }
     }
 
     /// Return the entry's output if it is valid for the given input.
-    fn lookup<In, Out>(&mut self, input: &In) -> Option<&Constrained<In::Constraint, Out>>
+    fn lookup<In>(&self, input: &In) -> Option<(&In::Constraint, &Out)>
     where
-        In: Input,
-        Out: Clone + 'static,
+        In: Input<Constraint = C>,
     {
-        let constrained: &Constrained<In::Constraint, Out> =
-            self.constrained.downcast_ref().expect("wrong entry type");
-
-        input.validate(&constrained.constraint).then(|| {
-            self.age = 0;
-            constrained
+        input.validate(&self.constraint).then(|| {
+            self.age.store(0, Ordering::Release);
+            (&self.constraint, &self.output)
         })
     }
 }
 
-/// Defines a constraint for a tracked type.
-#[derive(Clone)]
-pub struct Constraint<T>(RefCell<Vec<Call<T>>>);
-
 /// A call entry.
 #[derive(Clone)]
 struct Call<T> {
     args: T,
+    args_hash: u128,
     ret: u128,
     both: u128,
     mutable: bool,
 }
 
-impl<T: Hash + PartialEq> Constraint<T> {
+/// Defines a constraint for a tracked type.
+pub struct Constraint<T>(RwLock<Inner<T>>);
+
+#[derive(Clone)]
+struct Inner<T> {
+    /// The list of calls.
+    ///
+    /// Order matters here, as those are mutable & immutable calls.
+    calls: Vec<Call<T>>,
+    /// The hash of the arguments and index of the call.
+    ///
+    /// Order does not matter here, as those are immutable calls.
+    immutable: HashMap<u128, usize>,
```

Comment:
Is this really worth the extra complication, since it's only used for mutable constraints in the first place? In Typst, this would only affect `Tracer::inspected`.